### ISSUE TITLE
run0 doesn't load envvars from PAM

### ISSUE THREAD

**Issue by acid-bong** [systemd/systemd#34682]
### systemd version the issue has been seen with

256.4

### Used distribution

NixOS

### CPU architectures issue was seen on

None

### Component

systemd-run

### Expected behaviour you didn't see

Both interactive and non-interactive outputs should support non-ASCII symbols.

### Unexpected behaviour you saw

Only the interactive session displays Cyrillic properly, the non-interactive doesn't

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
~
❯ run0
==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-units ====
Authentication is required to manage system services or other units.
Authenticating as: acid
Password:
==== AUTHENTICATION COMPLETE ====
root in /home/acid
❯ ls
 ardour                        nvim-cmp.cast   src     Документы      'Рабочий стол'
 Documents                     olbackup        Sync    Загрузки        Шаблоны
 iso                           olbackup2       www     Изображения
 kernel-report                 org             бекуп   Музыка
'Me 4 Me [Kbt_ltV-9H8].webm'   prog            Видео   Общедоступные
root in /home/acid
❯
~ took 7s
❯ run0 ls
==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-units ====
Authentication is required to manage system services or other units.
Authenticating as: acid
Password:
==== AUTHENTICATION COMPLETE ====
 Documents
'Me 4 Me [Kbt_ltV-9H8].webm'
 Sync
 ardour
 iso
 kernel-report
 nvim-cmp.cast
 olbackup
 olbackup2
 org
 prog
 src
 www
''$'\320\222\320\270\320\264\320\265\320\276'
''$'\320\224\320\276\320\272\321\203\320\274\320\265\320\275\321\202\321\213'
''$'\320\227\320\260\320\263\321\200\321\203\320\267\320\272\320\270'
''$'\320\230\320\267\320\276\320\261\321\200\320\260\320\266\320\265\320\275\320\270\321\217'
''$'\320\234\321\203\320\267\321\213\320\272\320\260'
''$'\320\236\320\261\321\211\320\265\320\264\320\276\321\201\321\202\321\203\320\277\320\275\321\213\320\265'
...
```

**Comment by yuwata** [comment]
Is the string non-UTF ?

**Comment by yuwata** [comment]
At least, Japanese can be shown.
```sh
$ echo $LANG
C.utf8

$ ls /tmp/hoge
あああ  朝も早うからご苦労さん

$ build/run0 ls /tmp/hoge
あああ  朝も早うからご苦労さん

$ build/systemd-run --pty ls /tmp/hoge
Running as unit: run-u10907.service
Press ^] three times within 1s to disconnect TTY.
あああ	朝も早うからご苦労さん
```

**Comment by yuwata** [comment]
Also, the string copied from your report also shown correctly.
```sh
$ touch /tmp/hoge/бекуп

$ build/run0 ls /tmp/hoge
бекуп  あああ  朝も早うからご苦労さん
```

**Comment by acid-bong** [comment]
The final output seems to show it, but looks like there's a trouble with these three vars

```console
~
❯ locale
LANG=ru_RU.UTF-8
LC_CTYPE="ru_RU.UTF-8"
LC_NUMERIC="ru_RU.UTF-8"
LC_TIME="ru_RU.UTF-8"
LC_COLLATE="ru_RU.UTF-8"
LC_MONETARY="ru_RU.UTF-8"
LC_MESSAGES="ru_RU.UTF-8"
LC_PAPER="ru_RU.UTF-8"
LC_NAME="ru_RU.UTF-8"
LC_ADDRESS="ru_RU.UTF-8"
LC_TELEPHONE="ru_RU.UTF-8"
LC_MEASUREMENT="ru_RU.UTF-8"
LC_IDENTIFICATION="ru_RU.UTF-8"
LC_ALL=
~
❯ run0 !!
run0 locale
==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-units ====
Authentication is required to manage system services or other units.
Authenticating as: acid
Password:
==== AUTHENTICATION COMPLETE ====
/run/current-system/sw/bin/locale: Cannot set LC_CTYPE to default locale: No such file or directory
/run/current-system/sw/bin/locale: Cannot set LC_MESSAGES to default locale: No such file or directory
/run/current-system/sw/bin/locale: Cannot set LC_ALL to default locale: No such file or directory
LANG=ru_RU.UTF-8
LC_CTYPE="ru_RU.UTF-8"
LC_NUMERIC="ru_RU.UTF-8"
LC_TIME="ru_RU.UTF-8"
LC_COLLATE="ru_RU.UTF-8"
LC_MONETARY="ru_RU.UTF-8"
LC_MESSAGES="ru_RU.UTF-8"
LC_PAPER="ru_RU.UTF-8"
LC_NAME="ru_RU.UTF-8"
LC_ADDRESS="ru_RU.UTF-8"
LC_TELEPHONE="ru_RU.UTF-8"
LC_MEASUREMENT="ru_RU.UTF-8"
LC_IDENTIFICATION="ru_RU.UTF-8"
LC_ALL=
```

**Comment by acid-bong** [comment]
`LOCPATH` isn't set. Mount layout is the same, but with `run0` the lines are made of ASCII symbols instead of box-drawing ones:
```console
~
❯ lsblk
NAME      MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINTS
sda         8:0    0 447,1G  0 disk
├─sda1      8:1    0   512M  0 part  /boot
└─sda2      8:2    0 446,6G  0 part
  └─nixon 254:0    0 446,6G  0 crypt /nix/store
                                     /
sr0        11:0    1  1024M  0 rom
~
❯ run0 !!
run0 lsblk
==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-units ====
Authentication is required to manage system services or other units.
Authenticating as: acid
Password:
==== AUTHENTICATION COMPLETE ====
NAME      MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINTS
sda         8:0    0 447.1G  0 disk
|-sda1      8:1    0   512M  0 part  /boot
`-sda2      8:2    0 446.6G  0 part
  `-nixon 254:0    0 446.6G  0 crypt /nix/store
                                     /
sr0        11:0    1  1024M  0 rom
```
(`lsblk` for a shorter output)

**Comment by acid-bong** [comment]
@yuwata for context, what's your distro?

**Comment by yuwata** [comment]
> @yuwata for context, what's your distro?

Fedora 40
```
$ uname -a
Linux thinkpad-x280 6.10.6-200.fc40.x86_64 #1 SMP PREEMPT_DYNAMIC Mon Aug 19 14:09:30 UTC 2024 x86_64 GNU/Linux

$ localectl
System Locale: LANG=C.UTF-8
    VC Keymap: (unset)     
   X11 Layout: (unset)     

$ locale
LANG=C.utf8
LC_CTYPE="C.utf8"
LC_NUMERIC="C.utf8"
LC_TIME="C.utf8"
LC_COLLATE="C.utf8"
LC_MONETARY="C.utf8"
LC_MESSAGES="C.utf8"
LC_PAPER="C.utf8"
LC_NAME="C.utf8"
LC_ADDRESS="C.utf8"
LC_TELEPHONE="C.utf8"
LC_MEASUREMENT="C.utf8"
LC_IDENTIFICATION="C.utf8"
LC_ALL=

$ build/run0 locale
LANG=C.UTF-8
LC_CTYPE="C.UTF-8"
LC_NUMERIC="C.UTF-8"
LC_TIME="C.UTF-8"
LC_COLLATE="C.UTF-8"
LC_MONETARY="C.UTF-8"
LC_MESSAGES="C.UTF-8"
LC_PAPER="C.UTF-8"
LC_NAME="C.UTF-8"
LC_ADDRESS="C.UTF-8"
LC_TELEPHONE="C.UTF-8"
LC_MEASUREMENT="C.UTF-8"
LC_IDENTIFICATION="C.UTF-8"
LC_ALL=
```

**Comment by poettering** [comment]
Frankly I think this is a problem of NixOS, that needs to be dealt with in NixOS context, not here. Can you ask the NixOS folks for help on this, so that we can close this here?

**Comment by acid-bong** [comment]
There's one `run0`-specific detail: like `doas`, it doesn't seem to load envvars from `/etc/pam/environment`. The exact problem is that `run0` didn't load `LOCALE_ARCHIVE` envvar, which is necessary on NixOS for locale settings.

This is how `sudo` loads the non-interactive environment:
```console
❯ sudo printenv
[sudo] пароль для acid:
XAUTHORITY=/home/acid/.local/share/sx/xauthority
LANG=ru_RU.UTF-8
TERM=st-256color
DISPLAY=:1
TERMINFO_DIRS=/home/acid/.nix-profile/share/terminfo:/nix/profile/share/terminfo:/home/acid/.local/state/nix/profile/share/terminfo:/etc/profiles/per-user/acid/share/terminfo:/nix/var/nix/profiles/default/share/terminfo:/run/current-system/sw/share/terminfo
PATH=/run/wrappers/bin:/home/acid/.nix-profile/bin:/nix/profile/bin:/home/acid/.local/state/nix/profile/bin:/etc/profiles/per-user/acid/bin:/nix/var/nix/profiles/default/bin:/run/current-system/sw/bin
LS_COLORS=...
MAIL=/var/mail/root
LOGNAME=root
USER=root
HOME=/root
SHELL=/run/current-system/sw/bin/zsh
SUDO_COMMAND=/run/current-system/sw/bin/printenv
SUDO_USER=acid
SUDO_UID=1000
SUDO_GID=100
[-- NOT A PART OF THE OUTPUT; here start the envvars from PAM --]
GIO_EXTRA_MODULES=/nix/store/6xd7halb89mlp2c44i6pb5hpmhi5d135-dconf-0.40.0-lib/lib/gio/modules
GTK_A11Y=none
GTK_PATH=...
[...]
```
And the `run0 printenv` is the same as a couple of messages up there.

**Comment by poettering** [comment]
That seems to be a nixos problem. run0 does not bother with /etc/pam/environment, that's the job of pam_env.so. If you want that, enable it in the run0 PAM stack, but that has to be done by your distro or you personally, it's not someting we can configure/know from upstream.

hence, closing: please contact your distro, about enabling that PAM module, or not.

**Comment by acid-bong** [comment]
Welp, I once again tried making such a PAM config for `run0` (specifically with the `session required pam_env.so conffile=/etc/pam/environment` line) and it didn't work. It's definitely `run0`'s issue

**Comment by poettering** [comment]
OK, so this *is* indeed a systemd bug. he PAM stuff gets partially skipped in cases where it shouldn't be. Sorry for the confusion. Prepping a fix.

**Comment by poettering** [comment]
Fix pending here: #34880
